### ISSUE TITLE
python 3 failing

### ISSUE THREAD

**Issue by eyberg** [nanovms/nanos#901]
```
➜  t  ops load python_3.6.7 -n -c config.json
Downloading.. https://storage.googleapis.com/packagehub/python_3.6.7.tar.gz
 14.50 MiB / 14.50 MiB [=============================================================================================================================================================] 100.00% 9.32 MiB/s 1s
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-darwin.tar.gz
 588.33 KiB / 588.33 KiB [===========================================================================================================================================================] 100.00% 7.42 MiB/s 0s
[python3 -m http.server]
booting /Users/eyberg/.ops/images/python3.img ...
assigned: 10.0.2.15
Unhandled: 000000000000000d
General protection fault
interrupt: 000000000000000d
frame: 0000000101001600
error code: 0000000000000000
  rax: 0000000100549f48
  rbx: 000000010054a018
  rcx: 0000000100549f48
  rdx: 0000000100803000
  rsi: 000000010054a060
  rdi: 001d001e00400007
  rbp: 000000007fb84ef0
  rsp: 000000007fb84eb8
   r8: 0000000003000000
   r9: 0000000100803000
  r10: 00000000000001ff
  r11: 0000000000000000
  r12: 0000000103804000
  r13: 0000000103806000
  r14: 0000000000000004
  r15: 0000000100776140
  rip: 000000007f016f0d (notify_remove + 000000000000002d/0000000000000061)
flags: 0000000000000006
   ss: 0000000000000010
   cs: 0000000000000008
   ds: 0000000000000000
   es: 0000000000000000
   fs: 0000000000000000
   gs: 0000000000000000

frame trace:
000000007f018e3d        (poll_internal + 000000000000028d/00000000000008fb)
000000007f023feb        (syscall_debug + 000000000000018b/000000000000023c)
000000007f0002d2        (syscall_enter + 000000000000006a/0000000000000072)

stack trace:
000000007f017942        (free_epollfd + 0000000000000162/000000000000022b)
0000000000000000
0000000000000000
0000000103806000
00000000721f81a8
0000000103804000
0000000000000000
000000007fb84f80
000000007f018e3d        (poll_internal + 000000000000028d/00000000000008fb)
0000000103805050
000000010054a160
0000000000000000
0000000000000001
0000000000000000
0000000500000000
0000000103805000
000000007f04a668        (bootstrap_region + 0000000000000b28/0000000000200000)
000000010054a178
000000010054a168
00000000721f81a8
0000000100a02000
000000007f0194b0        (poll + 0000000000000000/0000000000000037)
0000000000000015
0000000100a02000

halt
exit status 255
➜  t  cat config.json
{
  "Args":["-m", "http.server"],
  "RunConfig": {
        "Ports": [8000]
  }
}
```

**Comment by ssudake21** [comment]
Please reenable Python3 test once this is resolved.

**Comment by x64k** [comment]
Will do! Sorry it's taking so long -- this is the most "fun" crash I've seen in a while now...

**Comment by x64k** [comment]
Quick update for anyone who's interested.

After a fun trip through Python's disassembled innards, I think I finally pinpointed the crashing code.

It looks like,  python's `freechildren` (which recursively does exactly what it sounds like it does, hence the very particular look of the stack trace) eventually runs into this situation:

```
unix_fault_page error: no vmap found for vaddr 0xfb8
Unhandled: 000000000000000e
Page fault
...
  rax: 0000000000000008
  rbx: 0000000000000fa0
...
  rdi: 000000007cffa860
  rbp: 0000000000000fa0
  rsp: 000000007cffd020
   r8: 0000000000000008
   r9: 000000007cffa805
  r10: 0000000004523665
  r11: 0000000000000246
  r12: 0000000000000000
  r13: 00000070130102f0
  r14: 000000007cffd330
  r15: 0000000000000101
  rip: 00000000043a240d
...
stack trace:
0000000000000fa0
00000070d434ffe8
0000000000000000
00000000043a2439
000000000004ffd8
00000070932d0e18
0000000000000000
00000000043a2439
...
```

It gets here because it actually runs into a node (at `00000070d434ffe8`) which really believes it has a child at address 0xfa0 (that's rbp, eventually -- the disassembly looks a little weird now because I've got a big fat printf in the function, normally the layout is a lot saner). It tries to look 0x18 bytes (at n_nchildren, i.e. as part of the `NCH(n)` call in the for loop to see if it got at the bottom of the tree of children) past that and, unsurprisingly, that's not an address that it can access...

`00000070d434ffe8` is suspiciously close to the boundary of a memory zone that got mremapped, so the leading hypothesis I have is that mremap is corrupting it somehow -- either the memory itself or the page mapping (at first I was tempted to think it's an identity mapping that didn't get removed but I don't think that's the case?). If not, I have a few other avenues that I want to look into -- this is just the proverbial horse that I think of when I hear hoofs (the zebras include race conditions, eww). But I think the fuzziest part is settled now.

**Comment by x64k** [comment]
OK, I think I got the root cause for this.

It seems that, in some cases, mremap is not able to correctly restore the virtual -> physical mapping when extending an already allocated portion. This is pretty easy to reproduce with a Python image (I suppose it just so happens that it fragments the memory space well enough). I managed to reproduce it with a trivial test program, too, but far less often (see the end of this message).

Please bear with me -- I know this is the kind of extraordinary claim that requires extraordinary evidence, so 80% of the reason why I'm commenting is to ask for second opinions, maybe I'm missing something here? (@wjhun , @mkhon , your names show up in the git log for mmap.c so I'd appreciate it if you could have a look at my analysis).

First, here's what happens on a Python test run.

Python's `freechildren` function crashes when it runs into a node that's obviously malformed. A node looks like this:

```
typedef struct _node {
    short               n_type;
    char                *n_str;
    int                 n_lineno;
    int                 n_col_offset;
    int                 n_nchildren;
    struct _node        *n_child;
} node;
```

nodes are part of a tree (I guess from the parser? I tried to look upwards but all it got me was a wasted day). They're freed recursively: for each node, we look at `n_nchildren`, which should be the length of `n_child`, and call the free function on each child. A node is freed when it has no children.

Now, at some point, we run into a tree that looks like this: 

```
1 0x000000704a04ffe8: 0x0000000000000001
1 0x000000704a04fff0: 0x0000007002540030
1 0x000000704a04fff8: 0x0000000000000000
1 0x000000704a050000: 0x0000000000000065
1 0x000000704a050008: 0x0000000000000000
```

This node is quite obviously malformed: it's supposed to have `0x65` children but `n_child` points to NULL. Just to make sure struct packing isn't fooling me, I dumped all this info from Python, too:

```
Node 0x704a04ffe8 has child 100 at bad address 0xfa0
...
-> Node = 0x704a04ffe8
...
    type: 0x704a04ffe8
...
    n_str: 0x704a04fff0
...
    n_lineno: 0x704a04fff8
...
    n_col_offset: 0x704a04fffc
...
    n_nchildren: 0x704a050000
...
    n_child: 0x704a050008
...
Node 0xfa0 will crash this.
```

(Ellipses correspond to log entries emitted by nanos because, well, printf prints by calling write etc.)

Sure enough, it crashes when attempting to work on the 100th noe, which just so happens to point at 0xfa0 (from this point on the crash is pretty obvious: it tries to inspect n_nchildren, so it attempts to look 0x18 bytes from the beginning of the node, attempting to read 0xfb0 and thus crashing with the log from above).

What's up with this node? Well, `0x704a04ffe8` just so happens to belong to a portion that had been previously been mmapped at some location, but has just been mremapped:

```
1 mremap
1 mremap: old_address 0x000000704cf00000, old_size 331776, new_size 659456, flags 0x1, new_address 0x000000704cf00000
1    new physical pages at 0x10800000, size 327680
1    old mapping at phys addr 0x1061d000, unmapping
1    mapping existing portion at 0x704a000000, phys addr 0x1061d000
1    mapping and zeroing new portion at 0x704a051000
```
So `0x704a04ffe8` is actually in a `0x51000`-long region starting at `0x704a000000` that used to be at `0x000000704cf00000`. `0x704a04ffe8` used to be at `0x704CF4FFE8`, and it should have been preserved. The first part of the mremapped memory region -- i.e. the "old" 0x51000 bytes -- should look the same at the new region, because they should be mapped to the same physical pages.

Dumping the memory reveals that the contents were not preserved. Here's what 0x704CF4FFE8 used to be:

```
1 0x000000704cf4ffe8: 0x000000000000000c
1 0x000000704cf4fff0: 0x00000070bfd8c2e0
1 0x000000704cf4fff8: 0x0000001e00000908
1 0x000000704cf50000: 0x0000000000000000
1 0x000000704cf50008: 0x0000000000000000

1 0x000000704cf50010: 0x0000000000000000
1 0x000000704cf50018: 0x0000000000000000
1 0x000000704cf50020: 0x0000000000000000
1 0x000000704cf50028: 0x0000000000000000
1 0x000000704cf50030: 0x0000000000000000
```
it looks nothing like what we see now:

```
1 0x000000704a04ffe8: 0x0000000000000001
1 0x000000704a04fff0: 0x0000007002540030
1 0x000000704a04fff8: 0x0000000000000000
1 0x000000704a050000: 0x0000000000000065
1 0x000000704a050008: 0x0000000000000000

1 0x000000704a050010: 0x0000000002686050
1 0x000000704a050018: 0x0000000002686050
1 0x000000704a050020: 0x0000000400000014
1 0x000000704a050028: 0x00000fd800000ff8
1 0x000000704a050030: 0x0000000000000137
```
This alone isn't much of an indication, of course -- it could be that the memory is corrupted, and that's actually what I thought at first, but I don't think that's the case.

Here's what the mappings of the relevant pages are when we enter mremap:

```
1 Hot mappings before remap:
1 0x000000704cf4f000-0x000000704cf50000:0x00000000107a0000-0x00000000107a1000
1 0x000000704cf50000-0x000000704cf51000:0x00000000107a9000-0x00000000107aa000

```
0x704CF4FFE8 (where our node used to be) resides on the first of these two pages. They are actually the last two pages of the 0x51000-long region that had originally been mmapped here:

```
1 mmap
1 mmap: target 0x0000000000000000, size 0x51000, prot 0x3, flags 0x22, fd -1, offset 0x0
1    anon target: new, 0x704cf00000, len: 0x51000 (given size: 0x51000)
```

Now mremap comes and tries to remap this region from 0x704a000000 (only it's 0xa1000 bytes long now). So it maps the old 0x51000-long portion from 0x704a000000, but here's what the mappings of the last two pages of this region look like now:

```
1 Hot mappings after remap:
1 0x000000704a04f000-0x000000704a050000:0x000000001066c000-0x000000001066d000
1 0x000000704a050000-0x000000704a051000:0x000000001066d000-0x000000001066e000
```

These two pages are probably peeking into another mmaped region's physical memory and, well, nothing good can come out of that.

It probably just so happens that it doesn't crash at the nodes prior to this point because they're all on the same page. But the "bad" node resides right on a page boundary. Half of its struct is from one node, but the other half belongs to God knows what else (probably a preamble to another set of nodes, judging by the look of the dump, but I won't bore you with it).

This is how I'm retrieving the mappings -- just to make sure I haven't missed something here:

```
    {
        void *paddr;
        u64 phys;
        paddr = (void *)vnew + old_size - (2 * 4096);
        phys  = physical_from_virtual(paddr);
        thread_log(current, "Hot mappings after remap:");
        thread_log(current, "%p-%p:%p-%p",
                   paddr, (paddr + 0x1000),
                   (void *)phys, (void *)(phys + 0x1000));

        paddr = (void *)vnew + old_size - (1 * 4096);
        phys  = physical_from_virtual(paddr);
        thread_log(current, "%p-%p:%p-%p",
                   paddr, (paddr + 0x1000),
                   (void *)phys, (void *)(phys + 0x1000));
    }

```

Now, just to test my hypothesis, I wrote a very trivial program:

[mmaptest.txt](https://github.com/nanovms/nanos/files/3408034/mmaptest.txt)

(Sorry for txt format but of course Github doesn't allow you to upload .c files, that's dangerous!)

It reproduces more or less the same pattern: it mmaps a private, anonymous region, then mremaps it at a larger size. The sizes are different so that it's easier to force mremapping to a different region (otherwise the original mapping remains unchanged -- I don't think it's relevant, but when I wrote the program I thought it might be). It then fills the last two pages with a recognizable pattern, and dumps the two pages so that they can be inspected.

Here's a successful run:

[mmaptest02-ok.txt](https://github.com/nanovms/nanos/files/3408048/mmaptest02-ok.txt)

Note how the two dumps of the last two pages are identical (as expected, after all), and the mappings are identical before and after the remapping:

1 Hot mappings before remap:
1 0x00000070938cf000-0x00000070938d0000:0x0000000002570000-0x0000000002571000
1 0x00000070938d0000-0x00000070938d1000:0x0000000002571000-0x0000000002572000
....
1 Hot mappings after remap:
1 0x00000070ca24f000-0x00000070ca250000:0x0000000002570000-0x0000000002571000
1 0x00000070ca250000-0x00000070ca251000:0x0000000002571000-0x0000000002572000

Here's an unsuccessful run:

[mmaptest01-fail-i-think.txt](https://github.com/nanovms/nanos/files/3408057/mmaptest01-fail-i-think.txt)

Note how the second dump looks pretty much like it's junk, but also look at the mappings:

```
1 Hot mappings before remap:
1 0x000000709614f000-0x0000007096150000:0x0000000002579000-0x000000000257a000
1 0x0000007096150000-0x0000007096151000:0x000000000257a000-0x000000000257b000

1 Hot mappings after remap:
1 0x000000709a94f000-0x000000709a950000:0x0000000002530000-0x0000000002531000
1 0x000000709a950000-0x000000709a951000:0x0000000002531000-0x0000000002532000
```

Fix is pending :).

I have a few relevant avenues for the fix, but I haven't looked at the code in mmap.c and page.c enough to speculate about it in public. My feeling is that mapping a large range of pages (i.e. the way mremap does by calling `map(vnew + old_size, dphys, dlen, mapflags, pages); `) doesn't work -- it just so happens that in the Python test, mappings are otherwise mostly done one page at a time. But I don't want to be a jerk and say oh, it's definitely <insert easy thing here>, even though <easy thing> is what I'd start with :).

**Comment by wjhun** [comment]
@x64k Sorry for the delayed response, and thank you for the excellent diagnosis. Indeed, mremap(2) is broken here as it was written back in the day before we had demand paging / faulting in of pages, so we were able to assume physical continuity. That's my fault for not catching that when I made those aforementioned updates.

Thanks again for finding this and giving such thorough notes. I commented on a proposed solution in #926.
